akka-persistence-testkit/src/main/scala/akka/persistence/testkit/InMemStorage.scala

```diff
@@ -0,0 +1,604 @@
+/*
+ * Copyright (C) 2018-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.testkit
+
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicReference
+import java.util.{ List => JList }
+
+import akka.util.ccompat.JavaConverters._
+import akka.actor.{ ActorSystem, ExtendedActorSystem, Extension, ExtensionId, ExtensionIdProvider }
+import akka.annotation.InternalApi
+import akka.persistence.testkit.ProcessingPolicy._
+import akka.persistence._
+import akka.serialization.{ Serialization, SerializationExtension, Serializers }
+import akka.persistence.testkit.scaladsl.{ PersistenceTestKit, SnapshotTestKit }
+
+import scala.annotation.tailrec
+import scala.collection.immutable
+import scala.util.{ Failure, Success, Try }
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+sealed trait InternalReprSupport[R] {
+
+  type InternalRepr
+
+  private[testkit] def toInternal(repr: R): InternalRepr
+
+  private[testkit] def toRepr(internal: InternalRepr): R
+
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+sealed trait InMemStorage[K, R] extends InternalReprSupport[R] {
+
+  import akka.persistence.testkit.Utils.JavaFuncConversions._
+
+  private val parallelism = Runtime.getRuntime.availableProcessors()
+
+  private final val eventsMap: ConcurrentHashMap[K, Vector[InternalRepr]] =
+    new ConcurrentHashMap()
```

Comment:
Given that this is a testkit without high performance requirements it would be possible to have a completely `synchronized` implementation and using ordinary Scala collections, but we can start like this and consider changing that if it becomes difficult to maintain or add more features.